commcare_connect/program/tests/test_helpers.py

```diff
@@ -0,0 +1,139 @@
+from datetime import timedelta
+
+import pytest
+from django_celery_beat.utils import now
+
+from commcare_connect.opportunity.models import VisitValidationStatus
+from commcare_connect.opportunity.tests.factories import AssessmentFactory, OpportunityAccessFactory, UserVisitFactory
+from commcare_connect.program.helpers import get_annotated_managed_opportunity
+from commcare_connect.program.tests.factories import ManagedOpportunityFactory, ProgramFactory
+from commcare_connect.users.tests.factories import OrganizationFactory, UserFactory
+
+
+class TestGetAnnotatedManagedOpportunity:
+    @pytest.fixture(autouse=True)
+    def setup(self, db):
+        self.program = ProgramFactory.create()
+        self.nm_org = OrganizationFactory.create()
+        self.opp = ManagedOpportunityFactory.create(program=self.program, organization=self.nm_org)
+
+    def create_user_with_access(self, visit_status=VisitValidationStatus.pending, passed_assessment=True):
+        user = UserFactory.create()
+        access = OpportunityAccessFactory.create(opportunity=self.opp, user=user, invited_date=now())
+        AssessmentFactory.create(opportunity=self.opp, user=user, opportunity_access=access, passed=passed_assessment)
+        UserVisitFactory.create(
+            user=user,
+            opportunity=self.opp,
+            status=visit_status,
+            opportunity_access=access,
+            visit_date=now() + timedelta(days=1),
+        )
+        return user
+
+    def test_basic_scenario(self):
+        for i in range(5):
+            self.create_user_with_access(
+                visit_status=VisitValidationStatus.pending if i < 3 else VisitValidationStatus.trial
+            )
+
+        opps = get_annotated_managed_opportunity(self.program)
+        assert len(opps) == 1
+        annotated_opp = opps[0]
+        assert annotated_opp.organization.slug == self.nm_org.slug
+        assert annotated_opp.workers_invited == 5
+        assert annotated_opp.workers_passing_assessment == 5
+        assert annotated_opp.workers_starting_delivery == 3
+        assert annotated_opp.percentage_conversion == 60.0
+
+    def test_empty_scenario(self):
+        opps = get_annotated_managed_opportunity(self.program)
```

Comment:
You could consider using parametrize, to combine some of the tests like below (not tested). Allows to be more DRY and concise.

```
    @pytest.mark.parametrize(
        "visit_statuses, passing_assessments, expected_invited, expected_passing, expected_delivery, expected_conversion",
        [
            ([VisitValidationStatus.pending, VisitValidationStatus.pending, VisitValidationStatus.trial], 
             [True, True, True], 5, 5, 3, 60.0),  # Basic scenario
            ([], [], 0, 0, 0, 0.0),  # Empty scenario
            ([VisitValidationStatus.pending] * 3, [True] * 3, 1, 1, 1, 100.0),  # Multiple visits scenario
            ([VisitValidationStatus.over_limit, VisitValidationStatus.trial], 
             [True, True], 2, 2, 0, 0.0),  # Excluded statuses
            ([VisitValidationStatus.pending, VisitValidationStatus.pending], 
             [False, True], 2, 1, 2, 100.0),  # Failed assessments
        ]
    )
    def test_scenarios(self, visit_statuses, passing_assessments, expected_invited, expected_passing, expected_delivery, expected_conversion):
        for i, visit_status in enumerate(visit_statuses):
            self.create_user_with_access(visit_status=visit_status, passed_assessment=passing_assessments[i])

        opps = get_annotated_managed_opportunity(self.program)
        assert len(opps) == 1
        annotated_opp = opps[0]
        assert annotated_opp.workers_invited == expected_invited
        assert annotated_opp.workers_passing_assessment == expected_passing
        assert annotated_opp.workers_starting_delivery == expected_delivery
        assert annotated_opp.percentage_conversion == expected_conversion
```

Reply:
Added [f87aa02](https://github.com/dimagi/commcare-connect/pull/391/commits/f87aa0223bd7fcd6b7a3857c1051e023e876068e)